Add tests for refs with relative uris

amosonn opened the pull request.

amosonn commented:
I could not find a way to port this to older versions, as they seem not to allow definitions in nested schemas. Is there a way to write a test with multiple, separate schemas?

gregsdennis commented:
@amosonn take a look at the `remotes` folder.  External schemas are stored here and can be referenced from the tests.

amosonn commented:
Added these tests, and also some for `dynamicRef`.

jdesrosiers reviewed (changes requested):
You can leave out the tests for `$recursiveRef` and `$dynamicRef`. `$recursiveRef` doesn't support any value other than an empty fragment (`"#"`), so that doesn't apply to what is being tested. `$dynamicRef` does support non-fragment-only URIs, but how that works is a little complicated. I'll be writing tests for `$dynamicRef` next week, and I'll make sure relative URI resolution is covered.

I've only skimmed this so far, but I'd be really surprised if relative URI resolution isn't covered somewhere already. I'll take a closer look later.

amosonn commented on remotes/ref-and-definitions.json:
```diff
@@ -0,0 +1,11 @@
+            }
+        }
+    },
+    "$ref": "#/definitions/inner"
```
I thought this was only true for verification keywords (e.g. `properties`), not for stuff like `definitions`. These test definitely pass on the client I'm using (JsonSchema.Net), meaning that `definitions` is found.

jdesrosiers replied:
It does work in many implementations, but not all. Whether an implementation supports it accidentally or on purpose, the ambiguity is reason enough to avoid it.

amosonn replied:
Yeah, I looked and the updated standard indeed writes "all other keywords". I changed it.

amosonn commented on tests/draft2019-09/recursiveRef.json:
```diff
@@ -352,5 +352,109 @@
+                            }
+                        }
+                    },
+                    "$recursiveRef": "#/$defs/inner"
```
I see, I will remove this. I thought that `$recursiveRef` behaved the same as `$ref` if there was not anchor, but missed that only `"#"` is defined.

amosonn commented on tests/draft2019-09/recursiveRef.json:
```diff
@@ -352,5 +352,109 @@
+                    "$recursiveRef": "#/$defs/inner"
+                }
+            },
+            "$recursiveRef": "schema2.json"
```
See above, will remove.

amosonn commented on tests/draft2020-12/dynamicRef.json:
```diff
@@ -253,5 +253,109 @@
+                            }
+                        }
+                    },
+                    "$dynamicRef": "#/$defs/inner"
```
Looking at the draft, it says:

> The value of the "$dynamicRef" property MUST be a string which is a URI-Reference. Resolved against the current URI base, it produces the URI used as the starting point for runtime resolution. This initial resolution is safe to perform on schema load.
> 
> If  the initially resolved starting point URI includes a fragment that was created by the "$dynamicAnchor" keyword, the initial URI MUST be replaced by the URI (including the fragment) for the outermost schema resource in the dynamic scope that defines an identically named fragment with "$dynamicAnchor".
> 
> Otherwise, its behavior is identical to "$ref", and no runtime resolution is needed. 

I didn't see any place which limits the URI to a plain identifier. I can remove this test if you find it better, but I think this is the correct behavior as defined by the current draft.

jdesrosiers replied:
I think you're right. It's nonsensical, but not technically wrong according to the spec. You can't expect any dynamic behavior because `$dynamicRef`s work with `$dynamicAnchor`s which can only be plain identifiers. Because the URI fragment can't match a `$dynamicAnchor`, it behaves like a `$ref` where that syntax is valid. That's terrible and not what the "identical to `$ref`" clause was intended for. You found a bizarre edge case that wasn't considered when the spec was written.

In any case, this test isn't supposed to be testing an edge case, it's supposed to be testing relative URI resolution. So, I'd rather see this test use `$dynamicRef` as it was intended. It will make more sense and it will age better.

jdesrosiers replied:
To elaborate a little more on the "identical to `$ref`" clause, that was meant as a fallback for when a `$dynamicAnchor` was not found that matches the URI fragment. It was not intended to allow `$dynamicRef` to be used in ways that weren't trying to do a dynamic reference.

`$dynamicRef` has some baggage it inherited from `$recursiveRef` that I think is the problem. Not everything that made sense for `$recursiveRef` makes sense for `$dynamicRef`. I'm trying to get it fixed for the next draft.

amosonn replied:
I think the best tests are for strange edge cases :) But if you think this an unimportant edge case which will get removed in the next draft, I can remove this; in that case, that draft should probably explicitly say that `$dynamicRef` is undefined when a matching `$dynamicAnchor` is not found, that it should only be a fragment, etc. (like `$recursiveRef` could only be the empty fragment, i.e. `#`)

As for a version which uses `$dynamicRef` "properly", I would suggest something like this:
```json
{
"$id": "schema1.json",
"properties": {
    "foo": {
        "$id": "schema2.json",
        "$dynamicAnchor": "foo"
        "$defs": {
            "inner": {
                "$dynamicAnchor": "inner"
                "properties": {
                    "bar": { "type": "string" }
                }
            }
        },
        "$dynamicRef": "#inner"
    },
    "$dynamicRef": "#foo"
}
```
And a similar one could be written with normal `$anchor`-s. However, at this point I feel this is too far removed from the original idea of this test, so I'm not sure whether it's relevant to include here. What do you think?

jdesrosiers replied:
> I think the best tests are for strange edge cases :) But if you think this an unimportant edge case which will get removed in the next draft, I can remove this

I don't think it needs a test, but that's not the point. The argument isn't that the edge case is not important or shouldn't be tested, it's that a test for an edge case shouldn't be lumped in with this test that has nothing to do with that edge case. A good test should test one thing and have one reason to fail. This test shouldn't fail because of a bug handling this edge case.

The alternate version looks reasonable to me.

amosonn replied:
Ok, as above, I'll make sure that the "split" test finds the relevant bugs, and then update.

amosonn replied:
I checked. My "compound" version above (the original) does indeed capture a bug that the "split" version doesn't. So I would prefer to add that test as-is. I can still add all the other tests written, which include separate ones for various `$dynamicRef`-s without `$dynamicAnchor`-s, and the one which uses them with an anchor.

amosonn replied:
This is taking too long and getting off-topic. I decided to completely remove the `$dynamicRef` tests from this PR, so it can be merged; once it lands, I'll open a new one where we can discuss solely those tests.

amosonn commented on tests/draft2020-12/dynamicRef.json:
```diff
@@ -253,5 +253,109 @@
+                    "$dynamicRef": "#/$defs/inner"
+                }
+            },
+            "$dynamicRef": "schema2.json"
```
See above, should behave like `$ref` if no `$dynamicAnchor` is present.

amosonn commented:
There are tests for relative uri resolution, I think, but not for relative uri resolution followed by `$defs` resolution. I got around to write these cases because they highlighted a bug in the client I'm using which wasn't covered by other tests from the suite, see https://github.com/gregsdennis/json-everything/issues/79 .

jdesrosiers commented on tests/draft2019-09/refRemote.json:
```diff
@@ -163,5 +163,28 @@
             }
         ]
+    },
+    {
+        "description": "remote ref with ref to defs",
+        "schema": {
+            "$id": "http://localhost:1234/schema1.json",
+            "$ref": "ref-and-defs.json"
+        },
+        "tests": [
+            {
+                "description": "invalid",
+                "data": {
+                    "bar": 1
+                },
+                "valid": false
+            },
+            {
+                "description": "valid",
+                "data": {
+                    "bar": "a"
+                },
+                "valid": true
+            }
+        ]
     }
```
I don't think this tests anything that the tests in "ref.json" don't. I'm not sure this is necessary.

amosonn replied:
I added them mostly because I already added them to draft 7 and 6 (which can't have the embedded `$defs`). But I would expect resolution of an external document to be differently implemented than of an embedded one, which justifies testing it as well?

jdesrosiers replied:
> draft 7 and 6 (which can't have the embedded `$defs`)

Um, yes they can.

> I would expect resolution of an external document to be differently implemented than of an embedded one

The way I've implemented it, there is no difference, but it is possible that some implementations might treat them differently. So, I take your point.

amosonn replied:
I tried those first (with embeded `definitions`), but they didn't work. I'll try it again, maybe this is also an implementation bug :)

amosonn replied:
Got it working, thanks for noting this!

jdesrosiers reviewed (changes requested):
I looked closer at the tests and they look fine except for some use of relative URIs for `$id`s at the top level of schemas.

I'm not sure "ref.json" is the right place for these tests. All other tests regarding embedding schemas with `$id` are in "refRemote.json". These should probably be there as well. I don't feel strongly about it tho if others disagree.

karenetheridge commented on tests/draft2019-09/ref.json:
```diff
@@ -430,5 +430,109 @@
+    {
+        "description": "refs with relative uris and defs",
+        "schema": {
+            "$id": "schema1.json",
```
I don't agree with that - a schema document can have a relative URI in its top-most $id.

jdesrosiers replied:
https://json-schema.org/draft/2020-12/json-schema-core.html#rfc.section.8.2.1
> If present, the value for this keyword MUST be a string, and MUST represent a valid URI-reference. This URI-reference SHOULD be normalized, and _MUST resolve to an absolute-URI_ (without a fragment).

karenetheridge replied:
`schema1.json` fits that criteria.

If schemeless or hostless $ids were prohibited, lots of other tests in the test suite would be incorrect as well.

amosonn replied:
With another view:

> ### 9.1.1. Initial Base URI
>
> Informatively, the initial base URI of a schema is the URI at which it was found, whether that was a network location, a local filesystem, or any other situation identifiable by a URI of any known scheme.

> ### 8.1.2 The "$id" Keyword
>
> Furthermore, in accordance with RFC 3986 section 5.1.2 regarding encapsulating entities, if an "$id" in a subschema is a relative URI-reference, the base URI for resolving that reference is the URI of the parent schema resource.
>
> If no parent schema object explicitly identifies itself as a resource with "$id", the base URI is that of the entire document, as established by the steps given in the previous section.

It does say so explicitly for subschemas, but I would interpret "resolve to an absolute-URI" as allowing a fall-back of a relative URI to using the "URI at which it was found" also for top-level ones.

I don't know if this is the best place to discuss this, but this is an important use-case for schemas which are hosted locally, therefore their absolute URI is not statically known, but they still refer to each other. It is useful for them to carry their relative URI, which can then be resolved relative to the base URI configured at the client.

jdesrosiers replied:
@karenetheridge How does `schema1.json` fit the definition of absolute-URI? https://tools.ietf.org/html/rfc3986#section-4.3
> absolute-URI  = scheme ":" hier-part [ "?" query ]

@amosonn It's not a fallback. They all apply. URLs resolve in a chain from the topmost context down. Retrieval URL > `$id` > embedded `$id` > `$ref`. If there is a retrieval URL, then a relative URI is ok for `$id` because it will resolve to an absolute URI. In contexts where there isn't a retrieval URL, an `$id` with a relative URI won't resolve to an absolute URI. Technically, it isn't against the spec to not have a retrieval URL or an `$id`, but that would mean that any `$ref`s that reference an external schema could not be relative because there isn't any URI to resolve it against.

Referencing local schemas is well supported. You can use retrieval URLs (`file://`) or use `$id` to identify your schemas. Remember that `$id`s don't need to be network accessible. They are just identifiers. You can make them up. Just put a scheme and domain in front of them,  `http://myapp.com/schema1`.

karenetheridge replied:
I don't think it's invalid for them to be schemeless. There are plenty of usecases where a schema is evaluated and there is no base URI provided at all (i.e. the base uri is '').

jdesrosiers replied:
@karenetheridge This discussion is going nowhere. Can we at least agree that changing this test to an absolute top level `$id` doesn't affect what is being covered by the test and that an absolute URI is a valid value for `$id`? That way there is nothing controversial and we can allow this PR to move forward. We can argue about what the spec says (and possibly what it should say) elsewhere.

karenetheridge replied:
sure. I did not mean to get in the middle of this particular PR.

amosonn replied:
I'm actually not quite sure that we can agree on this. Handling non-absolute URI-s is different than handling absolute ones, otherwise whether the spec allows this or not wouldn't be an issue. Part of what I wanted to test here is handling non-absolute ones, in combination with fragment-only ones (i.e. `$refs`), in particular. But if this is deemed too controversial, I can drop those for now. I will check that the original bugs are still found.

amosonn replied:
Ok, will update all top-level `$id`-s to be absolute.

amosonn commented on tests/draft2019-09/ref.json:
```diff
@@ -430,5 +430,109 @@
+                "foo": {
+                    "$id": "schema2.json",
+                    "$defs": {
+                        "inner": {
+                            "properties": {
+                                "bar": { "type": "string" }
+                            }
+                        }
```
Hmm, maybe you're right. It seems to me like it's nicer that the inner schema is used both for the top-level object and for the inner one; if it's a string then the property isn't checked at all. But if you think this is easier I can change it - I just have to verify that it still covers the original bug.

jdesrosiers replied:
Oops, when I wrote this I forgot that it was being used at the top and inner. Switching types would probably make the test more confusing even if the schema is simpler. So, I'm fine with it how it is.

amosonn commented:
Summary:

- I changed all the relative top-level `$id`-s to be absolute. I still think this is worth revisiting, but I agree that not in this PR.
- I removed the $dynamicRef tests from this PR completely, as that discussion was getting long. Once it lands, I'll open a new one where we can discuss solely those tests.

Anything else?

jdesrosiers commented on tests/draft2020-12/ref.json:
```diff
@@ -432,5 +432,109 @@
+    {
+        "description": "refs with relative uris and defs",
+        "schema": {
+            "$id": "http://example.com/schema1.json",
```
I notice that you use the same `$id` for each test. Technically, that shouldn't matter because each test _should_ be run in isolation, but I've noticed that there has been an effort in the past to use unique identifiers. It's fine with me, but we should make sure everyone is ok with it if we are going to allow duplicate `$id`s across tests.

karenetheridge replied:
Please use different $ids for each schema.  I *thought* we had a CI check in place for that, but maybe we just talked about it and no one has written it yet :)

jdesrosiers commented:
The "Test Suite Sanity Checking" build failure appears to be wrong. I think this is safe to merge.
